narwhal/primary/src/primary.rs

```diff
@@ -935,6 +970,19 @@ impl PrimaryToPrimary for PrimaryReceiverHandler {
                 None => continue,
             };
             if response.certificates.len() == request.max_items {
+                debug!(
+                    "Collected enough certificates (num={}, elapsed={}ms), returning.",
+                    response.certificates.len(),
+                    time_start.elapsed().as_millis(),
+                );
+                break;
+            }
+            if time_start.elapsed() >= FETCH_CERTIFICATES_MAX_HANDLER_TIME {
```

Comment:
If I understand correctly, this is a best effort approach?

Reply:
Yes, but each check only has a few local rocksdb reads so should be pretty close.